openshift/dynamic/client.py

```diff
@@ -454,13 +386,225 @@ def __getattr__(self, name):
         return partial(getattr(self.parent.client, name), self)
 
 
-class ResourceContainer(object):
+class Discoverer(object):
+    """
+        A convenient container for storing discovered API resources. Allows
+        easy searching and retrieval of specific resources.
+
+        Subclasses implement the abstract methods with different loading strategies.
+    """
+
+    def __init__(self, client):
+        self.client = client
+        self._load_server_info()
+
+    @abstractproperty
+    def api_groups(self):
+        pass
+
+    @abstractmethod
+    def get(self, prefix=None, group=None, api_version=None, kind=None, **kwargs):
+        pass
+
+    @abstractmethod
+    def search(self, prefix=None, group=None, api_version=None, kind=None, **kwargs):
+        pass
+
+    @property
+    def version(self):
+        return self.__version
+
+    def _load_server_info(self):
```

Comment:
I would recommend making this lazy as well, just make the request for version information when the version property is accessed. That will save us two requests per instantiation, since usually the version isn't used.

Reply:
Honestly I would go ahead and make this lazy just in this base implementation, there's really no need for it to be eager at all.

Reply:
I thought about this as well, but ultimately didn't make it lazy because for EagerDsicoverer, We call version to check if for openshift resources in the cluster in the `default_groups` function when setting up resources and for LazyDiscoverer I was calling it in __setup_resources.
e.g.
```
if self.version.get('openshift'):
    groups['oapi'] = { '': {
        'v1': self.ResourceGroup(True)
    }}
```

I'm going to look at getting rid of the group requests in LazyDiscoverer until they are actually needed like you mentioned in your other comment, then there may be some benefit.

Reply:
Yeah, that default groups logic is actually going to be broken in newer versions of openshift anyway, as I think they've removed the openshift version API entirely. Not really relevant to this PR though, just noting it.